### ISSUE TITLE
Private blockchain (i.e. Truffle develop/Ganache) gets out of sync with MetaMask

### ISSUE THREAD

**Issue by rhlsthrm** [MetaMask/metamask-extension#4361]
I am developing a dapp and testing with Metamask. Often through my development process, my Metamask gets out of sync with the blockchain, resulting in nonce errors. The workaround is to uninstall and reinstall Metamask, then re-import accounts, which is very time-consuming.

My reproduction steps for the issue are as follows:

* Connect Metamask to Custom RPC (http://localhost:9545 for `truffle develop`).
* Create transactions with Metamask (transactions are working).
* Re-migrate contracts by closing truffle develop console, reopening, and running `migrate --reset` (this preserves the contract deployment addresses).
* Refresh the page.
* Try to perform transactions. An error about nonce pops up.

I understand this is not behavior which will happen on a production blockchain, however, I rely on Metamask for development and it would be great to have a way to reset the account back to the initial state.

**Comment by danfinlay** [comment]
This workaround was made for just this problem:
https://metamask.helpscoutdocs.com/article/36-resetting-an-account

A longer term solution will require special communication between MetaMask and Ganache, which we hope to do eventually, but is unfortunately lower priority at the moment.

**Comment by mallapurbharat** [comment]
I'd like to reopen this issue, as I'm still seeing this issue (both on Chrome and Firefox) with the latest version (Metamask 4.8.0) on Ubuntu 16.04 . Ganache version is 1.2.0.

I've tried the following steps, but to no use.
1. Resetting the account
2. Switching networks and back again
3. Reinstallation of Metamask multiple times.

Though related, the unique problem is that my nonce in my local Ganache network is **8**, after deploying my contracts. However, Metamask is not picking up the network transaction activity and is still solidly at **0**. Hence I'm getting a nonce error (Metamask nonce: 0, Tx nonce: 8). Please see attached snapshots.

Please let me know asap if I should open up a new issue or whether you will reopen this issue. 

![ganache_8_transactions](https://user-images.githubusercontent.com/10878214/42671546-28248792-867e-11e8-897f-3f2213324c7e.png)
![js_console_error](https://user-images.githubusercontent.com/10878214/42671547-28519b10-867e-11e8-8ff7-cd74974193ad.png)

**Comment by mallapurbharat** [comment]
Ok... Did some troubleshooting. When I run Ganache 1.2.0 (and 1.2.1 which had some bug fixes), I faced this error of nonce mismatch. 
**Using Ganache 1.1.0, everything works fine**. So I doubt this is a bug in Metamask, though it appears that way. 

TLDR: Probably not a bug in Metamask. Resolved in my case by using older version of Ganache (1.1.0).
